File: solutions/heap/alternatives/692-top-k-frequent-words.js
/**
 * 692. Top K Frequent Words
 * Medium
 *
 * This problem demonstrates key concepts in Heap.
 *
 * SOLUTION EXPLANATION:
 *
 * INTUITION:
 * We need to find k most frequent words, with lexicographical ordering as tiebreaker.
 * A min heap of size k allows us to efficiently track the top k elements.
 *
 * APPROACH:
 * 1. **Count Frequencies**: Use hash map to count word frequencies
 * 2. **Build Min Heap**: Maintain heap of size k with custom comparator
 * 3. **Heap Comparator**: Compare by frequency (ascending), then by word (descending)
 * 4. **Extract Results**: Pop all from heap and reverse for correct order
 *
 * WHY THIS WORKS:
 * - Min heap keeps the k most frequent words
 * - Custom comparator ensures correct ordering
 * - Heap size k means O(log k) operations
 * - By using min heap, we can efficiently discard less frequent words
 *
 * TIME COMPLEXITY: O(n log k) where n is number of words
 * SPACE COMPLEXITY: O(n) for frequency map
 *
 * EXAMPLE WALKTHROUGH:
 * ```
 * Input: words = ["i","love","leetcode","i","love","coding"], k = 2
 * Step 1: Count: {i: 2, love: 2, leetcode: 1, coding: 1}
 * Step 2: Build heap of top k: ["i", "love"]
 * Step 3: Sort by frequency desc, then lexicographically
 * Output: ["i", "love"]
 * ```
 *
 * EDGE CASES:
 * - Single word
 * - All words have same frequency
 * - k equals number of unique words
 * - Words with same frequency (lexicographical order)
 */

/**
 * MinHeap implementation with custom comparator for [word, frequency] pairs
 * Comparison: First by frequency (ascending), then by word (descending for lexicographical)
 */
class MinHeap {
    constructor() {
        this.heap = [];
    }

    size() {
        return this.heap.length;
    }

    push(item) {
        this.heap.push(item);
        this.bubbleUp(this.heap.length - 1);
    }

    pop() {
        if (this.heap.length === 0) return null;
        if (this.heap.length === 1) return this.heap.pop();

        const top = this.heap[0];
        this.heap[0] = this.heap.pop();
        this.bubbleDown(0);
        return top;
    }

    // Custom comparator: returns true if a should be higher priority (smaller) than b
    compare(a, b) {
        // If frequencies are different, smaller frequency has higher priority (min heap)
        if (a[1] !== b[1]) {
            return a[1] < b[1];
        }
        // If frequencies are same, lexicographically larger word has higher priority
        // (so it gets removed first, keeping lexicographically smaller words)
        return a[0] > b[0];
    }

    bubbleUp(index) {
        while (index > 0) {
            const parentIndex = Math.floor((index - 1) / 2);
            if (!this.compare(this.heap[index], this.heap[parentIndex])) break;
            [this.heap[parentIndex], this.heap[index]] = [this.heap[index], this.heap[parentIndex]];
            index = parentIndex;
        }
    }

    bubbleDown(index) {
        while (true) {
            let smallest = index;
            const leftChild = 2 * index + 1;
            const rightChild = 2 * index + 2;

            if (leftChild < this.heap.length && this.compare(this.heap[leftChild], this.heap[smallest])) {
                smallest = leftChild;
            }
            if (rightChild < this.heap.length && this.compare(this.heap[rightChild], this.heap[smallest])) {
                smallest = rightChild;
            }
            if (smallest === index) break;

            [this.heap[index], this.heap[smallest]] = [this.heap[smallest], this.heap[index]];
            index = smallest;
        }
    }
}

/**
 * Main solution for Problem 692: Top K Frequent Words
 *
 * @param {string[]} words - Array of words
 * @param {number} k - Number of top frequent words to return
 * @return {string[]} - Top k frequent words in order
 *
 * Time Complexity: O(n log k)
 * Space Complexity: O(n)
 */
function solve(words, k) {
    if (!words || words.length === 0 || k === 0) {
        return [];
    }

    // Count word frequencies
    const freqMap = new Map();
    for (const word of words) {
        freqMap.set(word, (freqMap.get(word) || 0) + 1);
    }

    // Build min heap of size k
    const minHeap = new MinHeap();
    for (const [word, freq] of freqMap) {
        minHeap.push([word, freq]);
        if (minHeap.size() > k) {
            minHeap.pop();
        }
    }

    // Extract results and reverse (min heap gives us reverse order)
    const result = [];
    while (minHeap.size() > 0) {
        result.push(minHeap.pop()[0]);
    }

    return result.reverse();
}

/**
 * Test cases for Problem 692: Top K Frequent Words
 */
function testSolution() {
    console.log('Testing 692. Top K Frequent Words');

    // Test case 1: Basic functionality
    const result1 = solve(["i", "love", "leetcode", "i", "love", "coding"], 2);
    const expected1 = ["i", "love"];
    console.assert(JSON.stringify(result1) === JSON.stringify(expected1),
        `Test 1 failed: expected ${JSON.stringify(expected1)}, got ${JSON.stringify(result1)}`);

    // Test case 2: Lexicographical ordering
    const result2 = solve(["the", "day", "is", "sunny", "the", "the", "the", "sunny", "is", "is"], 4);
    const expected2 = ["the", "is", "sunny", "day"];
    console.assert(JSON.stringify(result2) === JSON.stringify(expected2),
        `Test 2 failed: expected ${JSON.stringify(expected2)}, got ${JSON.stringify(result2)}`);

    // Test case 3: Single word
    const result3 = solve(["a", "a", "a"], 1);
    const expected3 = ["a"];
    console.assert(JSON.stringify(result3) === JSON.stringify(expected3),
        `Test 3 failed: expected ${JSON.stringify(expected3)}, got ${JSON.stringify(result3)}`);

    // Test case 4: All unique words
    const result4 = solve(["a", "b", "c"], 2);
    const expected4 = ["a", "b"];
    console.assert(JSON.stringify(result4) === JSON.stringify(expected4),
        `Test 4 failed: expected ${JSON.stringify(expected4)}, got ${JSON.stringify(result4)}`);

    // Test case 5: k equals unique word count
    const result5 = solve(["i", "love", "coding"], 3);
    const expected5 = ["coding", "i", "love"];
    console.assert(JSON.stringify(result5) === JSON.stringify(expected5),
        `Test 5 failed: expected ${JSON.stringify(expected5)}, got ${JSON.stringify(result5)}`);

    console.log('All test cases passed for 692. Top K Frequent Words!');
}

/**
 * Example usage and demonstration
 */
function demonstrateSolution() {
    console.log('\n=== Problem 692. Top K Frequent Words ===');
    console.log('Category: Heap');
    console.log('Difficulty: Medium');
    console.log('');

    const words = ["i", "love", "leetcode", "i", "love", "coding"];
    const k = 2;
    console.log(`Input: words = [${words}], k = ${k}`);
    const result = solve(words, k);
    console.log(`Output: [${result}]`);

    testSolution();
}

// Run tests if this file is executed directly
if (require.main === module) {
    demonstrateSolution();
}

// Export for use in other modules
module.exports = {
    solve,
    testSolution,
    demonstrateSolution
};

/**
 * Additional Notes:
 * - This solution uses a min heap with custom comparator
 * - The comparator handles both frequency and lexicographical ordering
 * - Min heap of size k is more efficient than sorting all words
 * - Alternative: Use bucket sort for O(n) time if frequencies are bounded
 */

File: docs/solutions/heap/692-top-k-frequent-words.py
"""
# Difficulty: Medium

# 692. Top K Frequent Words

Given an array of strings words and an integer k, return the k most frequent strings.

Return the answer sorted by the frequency from highest to lowest. Sort the words with the same frequency by their lexicographical order.

**Example:**

<dl class="example-details">
<dt>Input:</dt>
<dd>[3, 2]</dd>
<dt>Output:</dt>
<dd>"Expected {expected}, got {result}"</dd>
<dt>Explanation:</dt>
<dd>Top k=2 frequent words in ['i','love','leetcode','i','love','coding'] are ['i','love']</dd>
</dl>

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
This problem combines frequency counting with sorting. We need to find the k most frequent words, but with a twist: when frequencies are equal, we sort lexicographically. A heap is perfect for this because we can maintain the top k elements efficiently while respecting both frequency and lexicographical ordering.

### APPROACH:
1. **Count frequencies**: Use HashMap to count word frequencies
2. **Use min-heap**: Maintain heap of size k with custom comparator
3. **Custom comparator**: Less frequent words first, then reverse lexicographical order
4. **Build result**: Extract from heap and reverse to get correct order

### WHY THIS WORKS:
- Min-heap keeps smallest elements at root, so we keep the k largest
- Custom comparator: (-frequency, word) ensures frequent words stay in heap
- For equal frequencies, lexicographically smaller words are "larger" in our heap
- Final result needs reversal because heap gives us reverse order

### EXAMPLE WALKTHROUGH:
```
Input: words = ["i","love","leetcode","i","love","coding"], k = 2
Frequencies: {"i": 2, "love": 2, "leetcode": 1, "coding": 1}
Heap process:
- Add ("i", 2): heap = [(-2, "i")]
- Add ("love", 2): heap = [(-2, "love"), (-2, "i")] (love > i lexicographically)
- Add ("leetcode", 1): heap = [(-1, "leetcode"), (-2, "i"), (-2, "love")]
- Remove leetcode: heap = [(-2, "love"), (-2, "i")]
- Add ("coding", 1): heap = [(-1, "coding"), (-2, "i"), (-2, "love")]
- Remove coding: heap = [(-2, "love"), (-2, "i")]
Result: ["i", "love"] (after reversing)
```

### TIME COMPLEXITY:
O(N log k)
Where N is number of words, k is the result size. Heap operations are O(log k).

### SPACE COMPLEXITY:
O(N + k)
O(N) for frequency map, O(k) for heap

### EDGE CASES:
- k equals number of unique words
- All words have same frequency (pure lexicographical sort)
- Single word repeated
- k = 1 with multiple candidates

</details>
"""

class Solution:
    def topKFrequent(self, words: list[str], k: int) -> list[str]:
        """
        Find top k frequent words using min-heap approach.

        Args:
            words: List of strings
            k: Number of top frequent words to return

        Returns:
            List of k most frequent words sorted by frequency (desc) then lexicographically

        Time Complexity: O(N log k) where N is number of words
        Space Complexity: O(N + k) for frequency map and heap
        """
        import heapq
        from collections import Counter

        # Count word frequencies
        freq_map = Counter(words)

        # Use min-heap with custom comparator
        # Store as (frequency, word) where we want:
        # - Higher frequency words to stay in heap
        # - For same frequency, lexicographically smaller words to stay
        heap = []

        for word, freq in freq_map.items():
            # Push (-freq, word) to simulate max-heap for frequency
            # For equal frequencies, Python's default string comparison handles lexicographical order
            heapq.heappush(heap, (freq, word))

            # Keep only k elements
            if len(heap) > k:
                heapq.heappop(heap)

        # Extract results in reverse order (heap gives smallest first)
        result = []
        while heap:
            freq, word = heapq.heappop(heap)
            result.append(word)

        # Reverse to get highest frequency first, and for same frequency, lexicographical order
        return result[::-1]

    def topKFrequentOptimal(self, words: list[str], k: int) -> list[str]:
        """
        Optimal solution using heap with proper comparator.

        Args:
            words: List of strings
            k: Number of top frequent words to return

        Returns:
            List of k most frequent words
        """
        import heapq
        from collections import Counter

        # Count frequencies
        freq_map = Counter(words)

        # Use heap with tuple (freq, word) for proper ordering
        # For min-heap: we want to keep larger frequencies and smaller lexicographical words
        # So we use (-freq, word) to achieve max-heap behavior for frequency
        heap = []

        for word, freq in freq_map.items():
            heapq.heappush(heap, (-freq, word))

        # Extract top k elements
        result = []
        for _ in range(k):
            freq, word = heapq.heappop(heap)
            result.append(word)

        return result

    def topKFrequentAlternative(self, words: list[str], k: int) -> list[str]:
        """
        Alternative solution using sorting.

        Args:
            words: List of strings
            k: Number of top frequent words

        Returns:
            List of k most frequent words

        Time Complexity: O(N log N) where N is unique words
        Space Complexity: O(N) for frequency map
        """
        from collections import Counter

        # Count frequencies
        freq_map = Counter(words)

        # Sort by frequency (descending) then lexicographically (ascending)
        sorted_words = sorted(freq_map.items(), key=lambda x: (-x[1], x[0]))

        # Return top k words
        return [word for word, freq in sorted_words[:k]]

    def topKFrequentHeapSize(self, words: list[str], k: int) -> list[str]:
        """
        Solution maintaining heap of exactly size k.

        Args:
            words: List of strings
            k: Number of top frequent words

        Returns:
            List of k most frequent words
        """
        import heapq
        from collections import Counter

        freq_map = Counter(words)

        # Min-heap of size k
        # Use (freq, reverse_word) to handle lexicographical ordering correctly
        heap = []

        for word, freq in freq_map.items():
            if len(heap) < k:
                # For min-heap: smaller elements first
                # We want to evict less frequent words, so use positive frequency
                # For same frequency, evict lexicographically larger words
                heapq.heappush(heap, (freq, word))
            else:
                # Check if current word should replace heap root
                if freq > heap[0][0] or (freq == heap[0][0] and word < heap[0][1]):
                    heapq.heapreplace(heap, (freq, word))

        # Sort result by frequency (desc) then lexicographically (asc)
        result = sorted(heap, key=lambda x: (-x[0], x[1]))
        return [word for freq, word in result]

def test_solution():
    """
    Test cases for the solution.
    """
    solution = Solution()

    # Test case 1: Basic case
    result = solution.topKFrequent([1, 2, 3], 2)
    expected = [3, 2]
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 2: Empty input
    result = solution.topKFrequent([], 0)
    expected = []
    assert result == expected, f"Expected {expected}, got {result}"

    print("All test cases passed!")

if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    print("=== 692. Top K Frequent Words ===")

    # Demonstrate various approaches
    test_cases = [
        (["i", "love", "leetcode", "i", "love", "coding"], 2),
        (["the", "day", "is", "sunny", "the", "the", "the", "sunny", "is", "is"], 4),
        (["a", "b", "c"], 2),
    ]

    for words, k in test_cases:
        print(f"\nInput: words={words}, k={k}")

        result1 = solution.topKFrequent(words, k)
        result2 = solution.topKFrequentOptimal(words, k)
        result3 = solution.topKFrequentAlternative(words, k)

        print(f"Heap approach:    {result1}")
        print(f"Optimal heap:     {result2}")
        print(f"Sorting approach: {result3}")

    # Show detailed example
    print(f"\nDetailed example:")
    words = ["i", "love", "leetcode", "i", "love", "coding"]
    k = 2
    print(f"Words: {words}")
    print(f"Frequencies: i:2, love:2, leetcode:1, coding:1")
    print(f"Top {k} frequent: {solution.topKFrequent(words, k)}")
    print(f"Reasoning: 'i' and 'love' both have frequency 2 (highest)")
    print(f"'i' comes before 'love' lexicographically")
